fuel-merkle/src/sparse/merkle_tree.rs

```diff
@@ -134,27 +133,22 @@ where
         if self.root_node().is_placeholder() {
             self.set_root_node(leaf_node);
         } else {
-            let (path_nodes, side_nodes) = self.path_set(leaf_node.clone())?;
+            let (path_nodes, side_nodes) = self.path_set(*key)?;
             self.update_with_path_set(&leaf_node, path_nodes.as_slice(), side_nodes.as_slice())?;
         }
 
         Ok(())
     }
 
-    pub fn delete(&mut self, key: &Bytes32, data: &[u8]) -> Result<(), MerkleTreeError<StorageError>> {
+    pub fn delete(&mut self, key: &Bytes32) -> Result<(), MerkleTreeError<StorageError>> {
         if self.root() == *Self::empty_root() {
             // The zero root signifies that all leaves are empty, including the
             // given key.
             return Ok(());
         }
 
-        let leaf_node = Node::create_leaf(key, data);
-        if let Some(primitive) = self.storage.get(&leaf_node.hash())? {
-            let primitive = primitive.into_owned();
-            let leaf_node: Node = primitive.try_into().map_err(MerkleTreeError::DeserializeError)?;
-            let (path_nodes, side_nodes): (Vec<Node>, Vec<Node>) = self.path_set(leaf_node.clone())?;
-            self.delete_with_path_set(&leaf_node, path_nodes.as_slice(), side_nodes.as_slice())?;
-        }
+        let (path_nodes, side_nodes): (Vec<Node>, Vec<Node>) = self.path_set(*key)?;
+        self.delete_with_path_set(key, path_nodes.as_slice(), side_nodes.as_slice())?;
```

Comment:
The problem here is that we are always deleting along this path, even if there was never anything added. This is problematic because we will revert nodes along the path that weren't actually changed, causing our root to be incorrect.